File: benchmarks/json_parse/JsonParse.java
public class JsonParse {
    static final String SRC = "{\"users\":[{\"id\":1,\"name\":\"Alice\",\"active\":true,\"scores\":[85,92,78]},{\"id\":2,\"name\":\"Bob\",\"active\":false,\"scores\":[91,88,95]},{\"id\":3,\"name\":\"Charlie\",\"active\":true,\"scores\":[76,84,90]}],\"metadata\":{\"version\":2,\"generated\":false,\"tags\":[\"benchmark\",\"json\",\"test\"],\"config\":{\"timeout\":30,\"retries\":3}}}";

    static class Parser {
        String src;
        int pos;

        Parser(String src) { this.src = src; }

        char peek() { return src.charAt(pos); }
        void advance() { pos++; }
        boolean eof() { return pos >= src.length(); }
        static boolean isDigit(int c) { return c >= '0' && c <= '9'; }
        static boolean isSpace(int c) { return c == ' ' || c == '\t' || c == '\n' || c == '\r'; }

        void skipWS() {
            while (!eof() && isSpace(peek())) advance();
        }

        long parseString() {
            advance();
            long n = 0;
            while (peek() != '"') { advance(); n++; }
            advance();
            return n;
        }

        long parseNumber() {
            long sign = 1;
            if (peek() == '-') { advance(); sign = -1; }
            long val = 0;
            while (!eof() && isDigit(peek())) {
                val = val * 10 + (peek() - '0');
                advance();
            }
            return sign * val;
        }

        long parseValue() {
            skipWS();
            if (eof()) return 0;
            int c = peek();
            if (c == '"') return parseString();
            if (c == 't') { advance(); advance(); advance(); advance(); return 1; }
            if (c == 'f') { advance(); advance(); advance(); advance(); advance(); return 0; }
            if (c == 'n') { advance(); advance(); advance(); advance(); return 0; }
            if (c == '-' || isDigit(c)) return parseNumber();
            if (c == '[') return parseArray();
            if (c == '{') return parseObject();
            return 0;
        }

        long parseArray() {
            advance(); skipWS();
            if (peek() == ']') { advance(); return 0; }
            long sum = 0;
            while (true) {
                sum += parseValue();
                skipWS();
                if (peek() == ',') advance();
                else break;
            }
            advance();
            return sum;
        }

        long parseObject() {
            advance(); skipWS();
            if (peek() == '}') { advance(); return 0; }
            long sum = 0;
            while (true) {
                skipWS();
                parseString();
                skipWS();
                advance();
                sum += parseValue();
                skipWS();
                if (peek() == ',') advance();
                else break;
            }
            advance();
            return sum;
        }
    }

    public static void main(String[] args) {
        Parser parser = new Parser(SRC);
        long total = 0;
        for (int i = 0; i < 100000; i++) {
            parser.pos = 0;
            total += parser.parseValue();
        }
        System.out.println(total);
    }
}

File: benchmarks/json_parse/json_parse.go
package main

import "fmt"

const src = `{"users":[{"id":1,"name":"Alice","active":true,"scores":[85,92,78]},{"id":2,"name":"Bob","active":false,"scores":[91,88,95]},{"id":3,"name":"Charlie","active":true,"scores":[76,84,90]}],"metadata":{"version":2,"generated":false,"tags":["benchmark","json","test"],"config":{"timeout":30,"retries":3}}}`

type Parser struct {
	src string
	pos int
}

func (p *Parser) peek() byte  { return p.src[p.pos] }
func (p *Parser) advance()     { p.pos++ }
func (p *Parser) eof() bool    { return p.pos >= len(p.src) }
func isDigit(c byte) bool      { return c >= '0' && c <= '9' }
func isSpace(c byte) bool      { return c == ' ' || c == '\t' || c == '\n' || c == '\r' }

func (p *Parser) skipWS() {
	for !p.eof() && isSpace(p.peek()) {
		p.advance()
	}
}

func (p *Parser) parseString() int64 {
	p.advance()
	var n int64 = 0
	for p.peek() != '"' {
		p.advance()
		n++
	}
	p.advance()
	return n
}

func (p *Parser) parseNumber() int64 {
	sign := int64(1)
	if p.peek() == '-' {
		p.advance()
		sign = -1
	}
	var val int64 = 0
	for !p.eof() && isDigit(p.peek()) {
		val = val*10 + int64(p.peek()-'0')
		p.advance()
	}
	return sign * val
}

func (p *Parser) parseValue() int64 {
	p.skipWS()
	if p.eof() {
		return 0
	}
	c := p.peek()
	switch {
	case c == '"':
		return p.parseString()
	case c == 't':
		p.advance()
		p.advance()
		p.advance()
		p.advance()
		return 1
	case c == 'f':
		p.advance()
		p.advance()
		p.advance()
		p.advance()
		p.advance()
		return 0
	case c == 'n':
		p.advance()
		p.advance()
		p.advance()
		p.advance()
		return 0
	case c == '-' || isDigit(c):
		return p.parseNumber()
	case c == '[':
		return p.parseArray()
	case c == '{':
		return p.parseObject()
	}
	return 0
}

func (p *Parser) parseArray() int64 {
	p.advance()
	p.skipWS()
	if p.peek() == ']' {
		p.advance()
		return 0
	}
	var sum int64 = 0
	for {
		sum += p.parseValue()
		p.skipWS()
		if p.peek() == ',' {
			p.advance()
		} else {
			break
		}
	}
	p.advance()
	return sum
}

func (p *Parser) parseObject() int64 {
	p.advance()
	p.skipWS()
	if p.peek() == '}' {
		p.advance()
		return 0
	}
	var sum int64 = 0
	for {
		p.skipWS()
		p.parseString()
		p.skipWS()
		p.advance()
		sum += p.parseValue()
		p.skipWS()
		if p.peek() == ',' {
			p.advance()
		} else {
			break
		}
	}
	p.advance()
	return sum
}

func main() {
	parser := Parser{src: src, pos: 0}
	var total int64 = 0
	for i := 0; i < 100000; i++ {
		parser.pos = 0
		total += parser.parseValue()
	}
	fmt.Println(total)
}
